### PR TITLE
Split connection logic for libp2p/webrtc peers

### PR DESCRIPTION
**Pull request by akoptelov**
- split connection/{incoming,outgoing} state machine parts to libp2p/webrtc
- allow multiple addresses for libp2p peers
- remove (redundant) peer id from dial opts

TODO:
- enable back kademlia-driven random peer selection
- initialize peers from initial peers (instead of doing it by kademlia)

### CODE REVIEW

**Review by binier — COMMENTED** [review]
Didn't review much but added couple comments.

@vlad9486 How does this pr go with your [state machine impl of libp2p](https://github.com/openmina/openmina/pull/178)?

File: node/src/effects.rs [hunk 1]
```diff
@@ -52,7 +53,8 @@ pub fn effects<S: Service>(store: &mut Store<S>, action: ActionWithMeta) {
 
             p2p_connection_timeouts(store, &meta);
 
-            store.dispatch(P2pConnectionOutgoingRandomInitAction {});
+            // TODO(akoptelov): reimplement this if needed
+            // store.dispatch(P2pConnectionOutgoingRandomInitAction {});
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply node/src/effects.rs:57]: That's why I have that TODO there )

File: node/src/effects.rs [hunk 2]
```diff
@@ -120,18 +122,43 @@ fn p2p_connection_timeouts<S: Service>(store: &mut Store<S>, meta: &ActionMeta)
...
+            match peer {
+                p2p::P2pPeerState::WebRTC(P2pWebRTCPeerState {
+                    status: P2pPeerStatus::Connecting(P2pConnectionState::Incoming(s)),
+                    ..
+                }) if s.is_timed_out(now) => Some((*peer_id, true, false)),
+                p2p::P2pPeerState::WebRTC(P2pWebRTCPeerState {
+                    status: P2pPeerStatus::Connecting(P2pConnectionState::Outgoing(s)),
+                    ..
+                }) if s.is_timed_out(now) => Some((*peer_id, true, true)),
+                // TODO(akoptelov): should libp2p layer be responsible for this?
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply node/src/effects.rs:134]: Yes, I agree. This was kinda a question to myself.

File: p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs [hunk 3]
```diff
@@ -5,70 +5,64 @@ use serde::{Deserialize, Serialize};
...
+};
 
-use super::P2pConnectionOutgoingInitOpts;
-
-#[derive(Serialize, Deserialize, Debug, Clone)]
-pub enum P2pConnectionOutgoingState {
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+pub enum P2pConnectionWebRTCOutgoingState {
+    #[default]
+    Default,
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs:16]: This is my attempt to have move as much logic as possible to e.g. outgoing connection reducer. 

I like your approach with having reducer as an associated function of the state, but before we can use it, we need that state. Upper level reducer just initializes the state with `default` and leaves the rest to the specific reducer.

File: p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs [hunk 4]
```diff
@@ -5,70 +5,64 @@ use serde::{Deserialize, Serialize};
...
     Init {
         time: Timestamp,
-        opts: P2pConnectionOutgoingInitOpts,
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs:16]: Because we already have address in the state. Why duplicate it?
```rust
pub struct P2pWebRTCPeerState {
    pub dial_opts: Option<SignalingMethod>,
    pub status: P2pWebRTCPeerStatus,
}
```

File: p2p/src/service_impl/webrtc_with_libp2p.rs [hunk 5]
```diff
@@ -40,27 +44,19 @@ pub trait P2pServiceWebrtcWithLibp2p: P2pServiceWebrtc {
...
 
-impl<T: P2pServiceWebrtcWithLibp2p> P2pConnectionService for T {
-    fn random_pick(
-        &mut self,
-        list: &[P2pConnectionOutgoingInitOpts],
-    ) -> P2pConnectionOutgoingInitOpts {
-        P2pServiceWebrtc::random_pick(self, list)
-    }
+impl<T: P2pServiceWebrtcWithLibp2p> P2pConnectionWebRTCService for T {
+    // fn random_pick(
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply p2p/src/service_impl/webrtc_with_libp2p.rs:51]: Yes, initial connection should be fixed, for libp2p as well.

File: p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (binier) [reply p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs:16]: Because it's a matter of when we set it. We set `dial_addr` on a successful connection, or if it's empty, if I'm not mistaken. Maybe that's no longer the case.

My knowledge of p2p is probably fairly outdated now, so if the above won't pose any problems, then it's fine I guess.

File: p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (binier) [reply p2p/src/connection/webrtc/outgoing/p2p_connection_webrtc_outgoing_state.rs:16]: Yeah that's an annoying part.

In that case, we should probably add invariant that after the reducer is done, the state is never in the Default variant.

**Comment by akoptelov** [comment]
> Didn't review much but added couple comments.

@binier The main things that I wasn't happy with, and why I even started this were:
- Having peers chosen by kademlia in the main state. As I mentioned earlier, these peers are first-class citizens, i.e. need to participate in gossipsub and RPC exchange. That mean as well that we should have their addresses persisted in the state, instead of having them only for connection.
- A mixture of libp2p and webrtc in the same actions enum, state enum, and reducer and effects as well. I had to guess and look into service to figure out if a particular reducer/effect is related to, and also how state will change.

Basically, the first bullet is a requirement, we barely can avoid this, especially when using @vlad9486 's p2p implementation. And the second one is for better readability and maintainability. Like Vlad suggested, later we can easily drop libp2p-style support as soon as there is a decision to use WebRTC.

So I would appreciate your opinion related to what I wrote above. Missing parts definitely should be fixed, and we can discuss how to approach this in the best way.

**Comment by binier** [comment]
> - Having peers chosen by kademlia in the main state. As I mentioned earlier, these peers are first-class citizens, i.e. need to participate in gossipsub and RPC exchange. That mean as well that we should have their addresses persisted in the state, instead of having them only for connection.
> - A mixture of libp2p and webrtc in the same actions enum, state enum, and reducer and effects as well. I had to guess and look into service to figure out if a particular reducer/effect is related to, and also how state will change.


@akoptelov I totally agree with both points.
